File: languages/python/bloock/entity/key/local_certificate_params.py
from __future__ import annotations

import bloock._bridge.proto.bloock_keys_entities_pb2 as proto
from bloock.entity.key.key_type import KeyType
from bloock.entity.key.subject_certificate_params import SubjectCertificateParams


class LocalCertificateParams:
    """
    Represents the parameters for generating a local certificate.
    """
    def __init__(
            self,
            key_type: KeyType,
            subject: SubjectCertificateParams,
            password: str,
            expiration: int = 0,
    ) -> None:
        """
        Constructs an LocalCertificateParams object with the specified parameters.
        :type expiration: object
        :type password: object
        :type subject: object
        :type key_type: object
        :rtype: object
        """
        self.key_type = key_type
        self.subject = subject
        self.password= password
        self.expiration = expiration

    @staticmethod
    def from_proto(certificate: proto.LocalCertificateParams) -> LocalCertificateParams:
        return LocalCertificateParams(
            key_type=KeyType.from_proto(certificate.key_type),
            subject=SubjectCertificateParams.from_proto(certificate.subject),
            password=certificate.password,
            expiration = certificate.expiration
        )

    def to_proto(self) -> proto.LocalCertificateParams:
        return proto.LocalCertificateParams(
            key_type=self.key_type.to_proto(),
            subject=self.subject.to_proto(),
            password=self.password,
            expiration = self.expiration
        )

File: languages/java/src/main/java/com/bloock/sdk/entity/key/LocalCertificateParams.java
package com.bloock.sdk.entity.key;

import com.bloock.sdk.bridge.proto.BloockKeysEntities;

/**
 * Represents the parameters for generating a local certificate.
 */
public class LocalCertificateParams {
  KeyType keyType;
  SubjectCertificateParams subject;
  String password;
  int expiration;

  /**
   * Constructs an LocalCertificateParams object with the specified parameters.
   * 
   * @param keyType
   * @param subject
   * @param password
   * @param expirationMonths
   */
  public LocalCertificateParams(
      KeyType keyType, SubjectCertificateParams subject, String password, int expirationMonths) {
    this.keyType = keyType;
    this.subject = subject;
    this.password = password;
    this.expiration = expirationMonths;
  }

  public static LocalCertificateParams fromProto(BloockKeysEntities.LocalCertificateParams params) {
    return new LocalCertificateParams(
        KeyType.fromProto(params.getKeyType()),
        new SubjectCertificateParams(
            params.getSubject().getCommonName(),
            params.getSubject().getOrganization(),
            params.getSubject().getOrganizationalUnit(),
            params.getSubject().getLocation(),
            params.getSubject().getState(),
            params.getSubject().getCountry()),
        params.getPassword(),
        params.getExpiration());
  }

  public BloockKeysEntities.LocalCertificateParams toProto() {
    BloockKeysEntities.LocalCertificateParams.Builder params = BloockKeysEntities.LocalCertificateParams.newBuilder()
        .setKeyType(this.keyType.toProto())
        .setSubject(this.subject.toProto())
        .setPassword(this.password)
        .setExpiration(this.expiration);

    return params.build();
  }
}
